File: Languages/C/Day025.c
// Languages - C - Day 25
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that converts a Decimal Number to Binary without using built-in conversion functions.


#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

void decimalToBinary(int n, char *binary)
{
    if (n == 0)
    {
        binary[0] = '0';
        binary[1] = '\0';
        return;
    }

    int isNegative = (n < 0);
    unsigned int num = abs(n);
    int i = 0;

    while (num > 0)
    {
        binary[i++] = (num % 2 == 0 ? '0' : '1');
        num /= 2;
    }

    if (isNegative)
    {
        binary[i++] = '-';
    }

    binary[i] = '\0';

    for (int j = 0; j < i / 2; ++j)
    {
        char temp = binary[j];
        binary[j] = binary[i - j - 1];
        binary[i - j - 1] = temp;
    }
}

int main()
{
    int input;
    char binary[CHAR_BIT * sizeof(int) + 2];
    printf("Enter a decimal number: ");

    if (scanf("%d", &input) != 1)
    {
        printf("Invalid input. Please enter an integer.\n");
        return 1;
    }

    decimalToBinary(input, binary);
    printf("Binary representation: %s\n", binary);

    return 0;
}

File: Languages/Rust/Day025.rs
// Languages - Rust - Day 25
// Foundational Concepts: Basics, syntax, loops, conditionals, functions, and variables.

// Create a function that converts a Decimal Number to Binary without using built-in conversion functions.

use std::io;

fn decimal_to_binary(mut number: i32) -> String {
    if number == 0 {
        return "0".to_string();
    }

    let mut binary = String::new();
    let is_negative = number < 0;
    number = number.abs();

    while number > 0 {
        binary.push_str(&(number % 2).to_string());
        number /= 2;
    }

    let binary = binary.chars().rev().collect::<String>();
    
    if is_negative {
        format!("-{}", binary)
    } else {
        binary
    }
}

fn main() {
    println!("Enter a decimal number:");
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    match input.trim().parse::<i32>() {
        Ok(number) => {
            let binary = decimal_to_binary(number);
            println!("Binary representation: {}", binary);
        }
        Err(_) => println!("Invalid input. Please enter a valid integer."),
    }
}
